### PR TITLE
feat(low-code cdk): add dynamic schema loader

### PR DESCRIPTION
**Pull request by lazebnyi**
## What
This PR add dynamic schema loader that can dynamically generate schema based on response.| 
Fixed: https://github.com/airbytehq/airbyte-internal-issues/issues/10782

## How 

Described schema, models and data classes for `DynamicSchemaLoader`, `SchemaTypeIdentifier` and `TypesPair`. Defined behavior for schema resolving. Integrated the dynamic schema handling into the existing framework.

## Release Notes

- **New Features**
  - Introduced three new components for enhanced dynamic schema handling: `TypesMap`, `SchemaTypeIdentifier`, and `DynamicSchemaLoader`.
  - Expanded type mappings in the manifest to support the new components.

- **Bug Fixes**
  - Improved error handling in the `DynamicSchemaLoader` for invalid keys and types.

- **Tests**
  - Added comprehensive unit tests for the `DynamicSchemaLoader` to validate functionality and error handling.

### CODE REVIEW

**Review by maxi297 — COMMENTED** [review]
Comments from sync discussion

@aldogonzalez8 can you have a look a this? You'll have to use this for google sheets

File: airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/parsers/model_to_component_factory.py:1590]: It looks like we have similar behavior in [DpathExtractor](https://github.com/airbytehq/airbyte-python-cdk/blob/cd1bd1c7119a31986b85c64cd88053cd4400dd1d/airbyte_cdk/sources/declarative/extractors/dpath_extractor.py#L62-L64) and [ParentStreamConfig](https://github.com/airbytehq/airbyte-python-cdk/blob/cd1bd1c7119a31986b85c64cd88053cd4400dd1d/airbyte_cdk/sources/declarative/partition_routers/substream_partition_router.py#L57-L60).

It might be a good idea to create a helper function as a follow-up. For now, I propose leaving it as is and revisiting this after merging this PR

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py]: updated

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/declarative_component_schema.yaml:1710]: Added

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/declarative_component_schema.yaml]: Updated

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/declarative_component_schema.yaml]: Updated

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/declarative_component_schema.yaml]: Removed this field from the implementation

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (aldogonzalez8) [inline airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py:104]: nit: I don't `parameters` being used. Is this to match some interface?

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (aldogonzalez8) [inline airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py:217]: Can you use different variable names for inner/outer variables to the list comprehension? Just for readability.

e.g. `calculated_path =` or `for current_extraction_path in` or whatever makes sense to you.

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (aldogonzalez8) [inline airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py:67]: nit: Should this be named `types_mapping` like in the yaml and model?

```
      types_mapping:
        type: array
        items:
          - "$ref": "#/definitions/TypesMap"
```

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py:104]: These need to align with the model's interpretation. In the model, we require this because parameters are passed from the top level to nested levels. In such cases, we need to explicitly specify the parameters, even if they are not directly used at that level, as they are utilized at the next level of the model in `SchemaTypeIdentifier`

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py]: updated

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py]: `path` updated to `node`

File: airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py
Comment (aldogonzalez8) [reply airbyte_cdk/sources/declarative/schema/dynamic_schema_loader.py:104]: Gotcha, thanks.

**Comment by aldogonzalez8** [comment]
The changes make sense to me. Just one thing, can we include a test with a manifest and the new dynamic schema loader setup? It would be nice to have like the ones in this other [PR](https://github.com/airbytehq/airbyte-python-cdk/pull/88).

File: airbyte_cdk/sources/declarative/models/declarative_component_schema.py
Comment (aldogonzalez8) [inline airbyte_cdk/sources/declarative/models/declarative_component_schema.py:1813]: I think you need to add DynamicSchemaLoader to schema_loader options [here](https://github.com/airbytehq/airbyte-python-cdk/blob/ed9a5e7aafb1ba4b9be04db70003e29d4c765de9/airbyte_cdk/sources/declarative/models/declarative_component_schema.py#L1612).

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (aldogonzalez8) [inline airbyte_cdk/sources/declarative/declarative_component_schema.yaml:1749]: Also, it must be added [here](https://github.com/airbytehq/airbyte-python-cdk/blob/main/airbyte_cdk/sources/declarative/declarative_component_schema.yaml#L1220-L1223) for what I can see.

**Review by maxi297 — COMMENTED** [review]
I think the only thing for me that would be missing is a very basic test that shows how it works in a manifest. Something similar to the `test_concurrent_declarative_source`. It doesn't need to be exhaustive but mostly to test that the discover method will return the dynamic stream

File: airbyte_cdk/sources/declarative/models/declarative_component_schema.py
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/models/declarative_component_schema.py:1830]: This is auto-generated file, so we don't need change manually anything here.

File: airbyte_cdk/sources/declarative/declarative_component_schema.yaml
Comment (lazebnyi) [reply airbyte_cdk/sources/declarative/declarative_component_schema.yaml:1750]: DynamicSchemaLoader has been added to DeclarativeStream schema

File: unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py
Comment (maxi297) [inline unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py:156]: It feels weird to have this comment here. We could have a `test_integration_dynamic_schema_loader.py` but I think I have a slight preference on having all those tests in this file

File: unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py
Comment (maxi297) [inline unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py:255]: We have this variable declared twice. Why is that? It feels like the declaration above should be removed

**Review by maxi297 — APPROVED** [review]
LGTM! I'm wondering if @aldogonzalez8 has used it in source-google-sheets and can approve this as well

File: unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py
Comment (aldogonzalez8) [reply unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py:156]: I feel the same for the imports, which should be moved above, but I think it was just a lousy copy-paste.

**Comment by aldogonzalez8** [comment]
> LGTM! I'm wondering if @aldogonzalez8 has used it in source-google-sheets and can approve this as well

I have some advance and I confirm I've been able to use the DynamicStreams and DynamicSchema for google-sheets with a [pre-dev cdk with minor changes](https://github.com/airbytehq/airbyte-python-cdk/compare/lazebnyi/add-dynamic-schema-loader...aldogonzalez8/add-dynamic-schema-loader-with-changes) that I think @lazebnyi had already added in last commits.

**Comment by aldogonzalez8** [comment]
> > LGTM! I'm wondering if @aldogonzalez8 has used it in source-google-sheets and can approve this as well
> 
> I have some advance and I confirm I've been able to use the DynamicStreams and DynamicSchema for google-sheets with a [pre-dev cdk with minor changes](https://github.com/airbytehq/airbyte-python-cdk/compare/lazebnyi/add-dynamic-schema-loader...aldogonzalez8/add-dynamic-schema-loader-with-changes) that I think @lazebnyi had already added in last commits.

Actualy is weird that the compare doesn't show me the latest changes in this PR but I assume is some bad cache or I did something wrong in the compare as I can see these options have already been added.

image


> This is auto-generated file, so we don't need change manually anything here.

@lazebnyi, what is the command/action that re-generated the file?

File: unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py
Comment (lazebnyi) [reply unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py:156]: These are artifacts from testing. My bad. They have been removed.

File: unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py
Comment (lazebnyi) [reply unit_tests/sources/declarative/schema/test_dynamic_schema_loader.py:255]: These are artifacts from testing. My bad. They have been removed.

**Pull request merged**
